src/utils/custom-actions.ts

```diff
@@ -0,0 +1,241 @@
+        errors.push(`Spotter-level custom actions cannot have position '${CustomActionsPosition.PRIMARY}'`);
+    }
+
+    // Check for primary action conflicts (this is a warning, not a validation
+    // failure)
+    if (position === CustomActionsPosition.PRIMARY) {
+        const existingPrimaryAction = primaryActionsPerTarget.get(targetType);
+        if (existingPrimaryAction) {
+            errors.push(`Multiple primary actions found for ${targetType.toLowerCase()}-level custom actions: '${existingPrimaryAction.name}' and '${action.name}'. Only the first action will be shown.`);
+        } else {
+            primaryActionsPerTarget.set(targetType, action);
+        }
+    }
+
+    // Validate metadata IDs
+    if (metadataIds) {
+        const invalidMetadataIds = Object.keys(metadataIds).filter(
+            (key) => !arrayIncludesString(config.allowedMetadataIds, key)
+        );
+        if (invalidMetadataIds.length > 0) {
+            errors.push(`Invalid metadata IDs for ${targetType.toLowerCase()}-level custom actions: ${invalidMetadataIds.join(', ')}`);
+        }
+    }
+
+    // Validate data model IDs
+    if (dataModelIds) {
+        const invalidDataModelIds = Object.keys(dataModelIds).filter(
+            (key) => !arrayIncludesString(config.allowedDataModelIds, key)
+        );
+        if (invalidDataModelIds.length > 0) {
+            errors.push(`Invalid data model IDs for ${targetType.toLowerCase()}-level custom actions: ${invalidDataModelIds.join(', ')}`);
+        }
+    }
+
+    // Validate allowed fields
+    const actionKeys = Object.keys(action);
+    const invalidFields = actionKeys.filter((key) => !arrayIncludesString(config.allowedFields, key));
+    if (invalidFields.length > 0) {
+        errors.push(`Invalid fields for ${targetType.toLowerCase()}-level custom actions: ${invalidFields.join(', ')}`);
+    }
+
+    return {
+        isValid: errors.length === 0,
+        reason: errors.length > 0 ? errors.join('; ') : '',
+    };
+};
+
+/**
+ * Validates basic action structure and required fields
+ * @param action - The action to validate
+ * @returns boolean indicating if the action has valid structure
+ * 
+ * @hidden
+ */
+const validateActionStructure = (action: any): boolean => {
+    if (!action || typeof action !== 'object') {
+        return false;
+    }
+
+    // Check for all missing required fields
+    const missingFields = ['id', 'name', 'target', 'position'].filter(field => !action[field]);
+    return missingFields.length === 0;
+};
+
+/**
+ * Checks for duplicate IDs among actions
+ * @param actions - Array of actions to check
+ * @returns Array of actions with unique IDs
+ * 
+ * @hidden
+ */
+const filterDuplicateIds = (actions: CustomAction[]): CustomAction[] => {
+    const idMap = new Map<string, CustomAction[]>();
+    actions.forEach((action) => {
+        const existing = idMap.get(action.id) || [];
+        existing.push(action);
+        idMap.set(action.id, existing);
```

Comment:
we use action.name as unique identifier within the app I believe. Ensure the unique-ness for the action name instead.